Flow reroute retry

rozdy opened the pull request.

niksv commented on src-java/reroute-topology/reroute-storm-topology/src/main/java/org/openkilda/wfm/topology/reroute/RerouteTopology.java:
```diff
@@ -65,26 +84,47 @@ public StormTopology createTopology() {
+        TimeWindowBolt timeWindowBolt = new TimeWindowBolt(topologyConfig.getRerouteThrottlingMinDelay(),
+                topologyConfig.getRerouteThrottlingMaxDelay());
+        // Time window bolt should use parallelism 1 to provide synchronisation for all reroute queue bolts
+        topologyBuilder.setBolt(BOLT_ID_TIME_WINDOW, timeWindowBolt, 1)
```
maybe use some constant for `1`?

rozdy replied:
I think comment is enough here.

niksv commented on src-java/reroute-topology/reroute-storm-topology/src/main/java/org/openkilda/wfm/topology/reroute/RerouteTopology.java:
```diff
@@ -53,9 +69,12 @@ public StormTopology createTopology() {
         TopologyBuilder topologyBuilder = new TopologyBuilder();
 
-        final Integer parallelism = topologyConfig.getParallelism();
+        final Integer parallelism = topologyConfig.getNewParallelism();
```
I'm a little bit worried about setting parallelism to 2 in almost all Bolts and spouts. We need to test it carefully

rozdy replied:
Design for this feature also include parallelism support. We are working on testing.

rozdy commented:
@surabujin 
> _[src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/fsm/reroute/actions/OnReceivedInstallResponseAction.java, line 110 at r2](https://reviewable.io/reviews/telstra/open-kilda/3295#-M36EQL306pVWdYDb9Sc:-M36EQL306pVWdYDb9Sd:btq3aw1) ([raw file](https://github.com/telstra/open-kilda/blob/29e49c863ff0c0b4cbfd768ef5a1bc681316a389/src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/fsm/reroute/actions/OnReceivedInstallResponseAction.java#L110)):_
> 
> > ```java
> >                 stateMachine.fire(Event.RULES_INSTALLED);
> >             } else {
> >                 String flowId = stateMachine.getFlowId();
> > ```
> 
> I do not remember any plans to get rid of per speaker request retry attempts. This change aimed on whole reroutes operation retries, not speaker requests retries. Have I missed something and we made a decision to get rid from speaker request retries?
>

This change is aimed to remove hotfix implementation of whole reroute retry. Per speaker retries are still aveilable.
 

> _[src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/fsm/reroute/actions/error/SetValidateRuleErrorAction.java, line 37 at r2](https://reviewable.io/reviews/telstra/open-kilda/3295#-M36I2Wy9CtnmLGr6j8j:-M36I2Wy9CtnmLGr6j8k:b-20ift3) ([raw file](https://github.com/telstra/open-kilda/blob/b6cee0dc1a29d2cc94a82f942f8680dd2342758a/src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/fsm/reroute/actions/error/SetValidateRuleErrorAction.java#L37)):_
> 
> > ```java
> >     @Override
> >     public void execute(State from, State to, Event event, FlowRerouteContext context, FlowRerouteFsm stateMachine) {
> > ```
> 
> (nip) too big (as for me) copy & paste factor. I would like to have some base class providing the method to collect all data and pass it into FSM instance. And set of inheritors responsible only for error message text.
> 

As for me inheritance here will add unnecessary complexity.

> _[src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/model/FlowRerouteFact.java, line 27 at r2](https://reviewable.io/reviews/telstra/open-kilda/3295#-M36IqVv12IOqy5TNuT2:-M36IqVv12IOqy5TNuT3:b-b4ezrg) ([raw file](https://github.com/telstra/open-kilda/blob/b6cee0dc1a29d2cc94a82f942f8680dd2342758a/src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/model/FlowRerouteFact.java#L27)):_
> 
> > ```java
> > @Value
> > public class FlowRerouteFact {
> > ```
> 
> This class was created to store info for reroute(internal) requests. Do we still need it?
> 

I think it's fine to use it as a storage for reroute request params with additional fields (key, command context) required by reroute service.

> _[src-java/reroute-topology/reroute-storm-topology/src/main/java/org/openkilda/wfm/topology/reroute/bolts/FlowRerouteQueueBolt.java, line 68 at r2](https://reviewable.io/reviews/telstra/open-kilda/3295#-M36sWk04K9ywPjqnERY:-M36sWk04K9ywPjqnERZ:b5iqrmp) ([raw file](https://github.com/telstra/open-kilda/blob/b6cee0dc1a29d2cc94a82f942f8680dd2342758a/src-java/reroute-topology/reroute-storm-topology/src/main/java/org/openkilda/wfm/topology/reroute/bolts/FlowRerouteQueueBolt.java#L68)):_
> 
> > ```java
> >             FlowThrottlingData throttlingData;
> >             switch (tuple.getSourceStreamId()) {
> >                 case STREAM_REROUTE_RESULT_ID:
> > ```
> 
> (nip) perhaps it will be simpler to use 1 stream and build routing on the event type (most part of our code work in this way). I don't like both of these approaches, I can't say which one worse.
> 

I need to use streams here to differ automatic reroute requests and manual. They have the same payload type.

> _[src-java/reroute-topology/reroute-storm-topology/src/main/java/org/openkilda/wfm/topology/reroute/bolts/TimeWindowBolt.java, line 73 at r2](https://reviewable.io/reviews/telstra/open-kilda/3295#-M37-AdqDkvUUQQQsOBr:-M37-Adr02zqjFfeaUY-:b9lyqta) ([raw file](https://github.com/telstra/open-kilda/blob/b6cee0dc1a29d2cc94a82f942f8680dd2342758a/src-java/reroute-topology/reroute-storm-topology/src/main/java/org/openkilda/wfm/topology/reroute/bolts/TimeWindowBolt.java#L73)):_
> 
> > ```java
> >         super.init();
> >         // Imitate flush window event on bolt startup to prevent data loss when this bolt is restarted
> >         getOutput().emit(new Values(new CommandContext()));
> > ```
> 
> Does the storm make any promises regarding readiness to transfer tuples at this stage?

Yep, according to storm docs `Tuples can be emitted at anytime from the bolt -- in the prepare, execute, or cleanup methods, or even asynchronously in another thread.` https://storm.apache.org/releases/current/Tutorial.html


All other comments are fixed.

rozdy commented:
@surabujin 
> _[src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/fsm/reroute/FlowRerouteFsm.java, line 158 at r5](https://reviewable.io/reviews/telstra/open-kilda/3295#-M3p1XFoAKtXqC5g1nvs:-M3p1XFoAKtXqC5g1nvt:b-yvzx9r) ([raw file](https://github.com/telstra/open-kilda/blob/572336e16b243b39117b328bbe914c37303771a1/src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/fsm/reroute/FlowRerouteFsm.java#L158)):_
> 
> > ```java
> >     public void setRerouteError(RerouteError rerouteError) {
> >         if (this.rerouteError == null) {
> >             this.rerouteError = rerouteError;
> > ```
> 
> Should we log this error (especially if `this.rerouteError != null`)?

I think errors should be logged where they appears but not here.

> _[src-java/reroute-topology/reroute-storm-topology/src/main/java/org/openkilda/wfm/topology/reroute/bolts/FlowRerouteQueueBolt.java, line 78 at r5](https://reviewable.io/reviews/telstra/open-kilda/3295#-M3pDRUw9YdNWVw8rCZe:-M3pDRUxCPXeJzxGZeNG:b-t1bqcr) ([raw file](https://github.com/telstra/open-kilda/blob/572336e16b243b39117b328bbe914c37303771a1/src-java/reroute-topology/reroute-storm-topology/src/main/java/org/openkilda/wfm/topology/reroute/bolts/FlowRerouteQueueBolt.java#L78)):_
> 
> > ```java
> >             case STREAM_REROUTE_RESULT_ID:
> >                 RerouteResultInfoData rerouteResultInfoData =
> >                         (RerouteResultInfoData) tuple.getValueByField(RerouteBolt.REROUTE_RESULT_FIELD);
> > ```
> 
> Why not `pullValue`?

Fixed.

> 
> _[src-java/reroute-topology/reroute-storm-topology/src/main/java/org/openkilda/wfm/topology/reroute/service/RerouteQueueService.java, line 140 at r5](https://reviewable.io/reviews/telstra/open-kilda/3295#-M3pG3Fm03ushgc_voiT:-M3pG3Fm03ushgc_voiU:be2rep3) ([raw file](https://github.com/telstra/open-kilda/blob/572336e16b243b39117b328bbe914c37303771a1/src-java/reroute-topology/reroute-storm-topology/src/main/java/org/openkilda/wfm/topology/reroute/service/RerouteQueueService.java#L140)):_
> 
> > ```java
> >         Map<String, FlowThrottlingData> requestsToSend = new HashMap<>();
> >         reroutes.forEach((flowId, rerouteQueue) -> {
> >             FlowThrottlingData flowThrottlingData = rerouteQueue.flushThrottling();
> > ```
> 
> Perhaps we should return `Optional` here.

Fixed.

> 
> _[src-java/reroute-topology/reroute-storm-topology/src/main/java/org/openkilda/wfm/topology/reroute/service/RerouteQueueService.java, line 149 at r5](https://reviewable.io/reviews/telstra/open-kilda/3295#-M3pGcCl7ElhPwkkFqxN:-M3pGcCl7ElhPwkkFqxO:b-5jh3h0) ([raw file](https://github.com/telstra/open-kilda/blob/572336e16b243b39117b328bbe914c37303771a1/src-java/reroute-topology/reroute-storm-topology/src/main/java/org/openkilda/wfm/topology/reroute/service/RerouteQueueService.java#L149)):_
> 
> > ```java
> >                 .sorted(Map.Entry.comparingByValue(new FlowPriorityComparator()))
> >                 .forEach(es -> {
> >                     String flowId = es.getKey();
> > ```
> 
> Do the `FlowThrottlingData` contain `flowId` (perhaps it should)?

No it doesn't. It may lead to strange situations with `FlowThrottlingData` for different flows in one queue so I decided to store flow id separately.
